tests/small_numbers/float32.ml

```diff
@@ -0,0 +1,267 @@
+
+    let via_f32 = Float32.to_int (Float32.of_float f) in
+    let via_int32 = Float.to_int via_int32 in
+    eq via_f32 via_int32
+  );
+  Int.check_ints (fun i ->
+    let via_f32 = Float32.to_float (Float32.of_int i) in
+    let via_f64 = Int32.float_of_bits (Int32.bits_of_float (Float.of_int i)) in
+    eqf via_f32 via_f64
+  )
+;;
+
+let () = (* Arithmetic *)
+  let assert_same a b =
+    assert (a = b || (CFloat32.is_nan a && CFloat32.is_nan b))
+  in
+  CFloat32.check_float32s (fun f1 f2 ->
+    assert_same (CFloat32.add f1 f2) (Float32.add f1 f2);
+    assert_same (CFloat32.sub f1 f2) (Float32.sub f1 f2);
+    assert_same (CFloat32.mul f1 f2) (Float32.mul f1 f2);
+    assert_same (CFloat32.div f1 f2) (Float32.div f1 f2);
+    assert_same (CFloat32.abs f1) (Float32.abs f1);
+    assert_same (CFloat32.neg f1) (Float32.neg f1)
+  )
+;;
+
+let () = (* Compare *)
+  let run compare =
+    CFloat32.check_float32s (fun f1 f2 ->
+      let no_nan = (not (CFloat32.is_nan f1)) && (not (CFloat32.is_nan f2)) in
+      if no_nan then (
+        (match Float32.compare f1 f2 with
+        | -1 -> assert (CFloat32.lt f1 f2)
+        | 0 -> assert (CFloat32.eq f1 f2)
+        | 1 -> assert (CFloat32.gt f1 f2)
+        | _ -> assert false);
+        let f2, f1 = f1, f2 in
+        (match Float32.compare f1 f2 with
+        | -1 -> assert (CFloat32.lt f1 f2)
+        | 0 -> assert (CFloat32.eq f1 f2)
+        | 1 -> assert (CFloat32.gt f1 f2)
+        | _ -> assert false);
+      ) else (
+        assert (CFloat32.neq f1 f2);
+        assert (CFloat32.neq f2 f1);
+        assert (CFloat32.nlt f1 f2);
+        assert (CFloat32.nle f1 f2);
+        assert (CFloat32.ngt f1 f2);
+        assert (CFloat32.nge f1 f2);
+        if CFloat32.is_nan f1 && CFloat32.is_nan f2 then (
+          assert (Float32.compare f1 f2 = 0)
+        );
+        if CFloat32.is_nan f1 && not (CFloat32.is_nan f2) then (
+          assert (Float32.compare f1 f2 = -1);
+          assert (Float32.compare f2 f1 = 1)
+        );
+        let f2, f1 = f1, f2 in
+        if CFloat32.is_nan f1 && not (CFloat32.is_nan f2) then (
+          assert (Float32.compare f1 f2 = -1);
+          assert (Float32.compare f2 f1 = 1)
+        )
+      )
+    )
+  in
+  (run [@inlined]) Float32.compare;
+  (run [@inlined never]) (fun l r -> Poly.compare (Sys.opaque_identity l) (Sys.opaque_identity r))
+;;
+
+let () = (* Hash *)
+  CFloat32.check_float32s (fun f _ ->
+      let h = Poly.hash f in
+      let bits =
+        if f = Float32.of_float 0.0 then 0l
+        else if CFloat32.is_nan f then 0x7F800001l
+        else CFloat32.bits_to_int f
+      in
+      assert (h = Poly.hash bits)
+  )
+;;
+
+let () = (* Marshal *)
+  CFloat32.check_float32s (fun l r ->
+    let b = Poly.to_bytes (Sys.opaque_identity l, Sys.opaque_identity r) [] in
+    let (readl, readr) : float32 * float32 = Poly.from_bytes_unsafe b 0 in
+    assert (Float32.compare l readl = 0);
+    assert (Float32.compare r readr = 0)
```

Comment:
(I wonder whether this should use the bit patterns.)

Reply:
Yep